elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,448 @@
+   * Define a custom element, `<dash-treeitem>`, and associate it with the
+   * `DashTreeItem` class.
+   */
+  window.customElements.define('dash-treeitem', DashTreeItem);
+
+  /**
+   * `DashTreeGroup` is a simple container that holds the children of a
+   * `DashTreeItem` parent node.
+   */
+  class DashTreeGroup extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      this.setAttribute('role', 'group');
+    }
+  }
+
+  /**
+   * Define a custom element, `<dash-treegroup>`, and associate it with the
+   * `DashTreeGroup` class.
+   */
+  window.customElements.define('dash-treegroup', DashTreeGroup);
+
+  /**
+   * `DashTree` is responsible for handling user input and updating the
+   * expanded/collapsed and selected state for its children. It also manages
+   * the currently active child using `aria-activedescendant`.
+   */
+  class DashTree extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tree');
+      this.setAttribute('tabindex', 0);
+
+      // Before the elements starts booting, it waits for both
+      // the `<dash-treeitem>` and `<dash-treegroup>` to load.
+      Promise.all([
+        customElements.whenDefined('dash-treeitem'),
+        customElements.whenDefined('dash-treegroup'),
+      ]).then(_ => {
+        // Acquire all treeitems inside the element
+        const treeItems = this._allTreeItems();
+        // If there are no treeItems, then the tree is empty. Abort.
+        if (treeItems.length === 0) return;
+
+        // The element checks if any child has been marked as selected.
+        // If so, it will mark it as the current `aria-activedescendant`
+        // and mark it with a `.selected` class.
+        const selectedTreeItem =
+          treeItems.find(treeItem =>
+            treeItem.getAttribute('aria-selected') === 'true');
+
+        if (selectedTreeItem) {
+          this._focusTreeItem(selectedTreeItem);
+          this._selectTreeItem();
+        }
+      });
+    }
+
+    /**
+     * Returns a list of visible `DashTreeItem` elements. This is useful when
+     * the user wants to try to move to the next or previous item in the list.
+     * If an item is a child of a parent who is set to `aria-expanded=false`
+     * then it is considered invisible and is not added to the list.
+     */
+    _allTreeItems() {
+      const treeItems = [];
+      // A recursive function that visits every child and builds a list
+      // This produces similar results to calling querySelectorAll,
+      // but allows for filtering of the children based on whether or not
+      // their parent is currently visible.
+      function findTreeItems(node, isVisible) {
```

Comment:
Do you still need the `isVisible` parameter at all? You start with `true` and never change it.